<a href="https://colab.research.google.com/github/HarshavardhanBhupathi/Data-science/blob/main/Creditcard_fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from sklearn.utils import resample
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import auc
from sklearn.metrics import plot_precision_recall_curve
from sklearn.metrics import accuracy_score



In [ ]:
cc=pd.read_csv("/home/creditcard.csv")

In [ ]:
cc.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0     0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62    0.0
1     0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69    0.0
2     1 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66    0.0
3     1 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50    0.0
4     2 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99    0.0

[5 rows x 31 columns]

In [ ]:
cc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93181 entries, 0 to 93180
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    93181 non-null  int64  
 1   V1      93181 non-null  float64
 2   V2      93181 non-null  float64
 3   V3      93181 non-null  float64
 4   V4      93181 non-null  float64
 5   V5      93181 non-null  float64
 6   V6      93181 non-null  float64
 7   V7      93181 non-null  float64
 8   V8      93181 non-null  float64
 9   V9      93181 non-null  float64
 10  V10     93181 non-null  float64
 11  V11     93181 non-null  float64
 12  V12     93181 non-null  float64
 13  V13     93180 non-null  float64
 14  V14     93180 non-null  float64
 15  V15     93180 non-null  float64
 16  V16     93180 non-null  float64
 17  V17     93180 non-null  float64
 18  V18     93180 non-null  float64
 19  V19     93180 non-null  float64
 20  V20     93180 non-null  float64
 21  V21     93180 non-null  float64
 22

In [ ]:
cc.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [ ]:
cc.nunique

<bound method DataFrame.nunique of         Time        V1        V2        V3  ...       V27       V28  Amount  Class
0          0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62    0.0
1          0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69    0.0
2          1 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66    0.0
3          1 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50    0.0
4          2 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99    0.0
...      ...       ...       ...       ...  ...       ...       ...     ...    ...
93176  64282  1.236457  0.214973  0.637424  ... -0.032589  0.021078    1.98    0.0
93177  64282  1.156176 -0.319928  0.795544  ... -0.010598  0.018465   29.39    0.0
93178  64283 -0.374903  0.013912  0.888019  ...  0.106798 -0.027620   30.00    0.0
93179  64283 -0.508209 -0.102169  2.060352  ... -0.092805 -0.177516   29.90    0.0
93180  64283 -0.537899  0.550337  0.106653  ...     

In [ ]:
cc.Time.unique

<bound method Series.unique of 0            0
1            0
2            1
3            1
4            2
         ...  
93176    64282
93177    64282
93178    64283
93179    64283
93180    64283
Name: Time, Length: 93181, dtype: int64>

In [ ]:
cc.shape

(93181, 31)

93181 rows is still too large for SVM algorithm.lets downsize the data using resample from sklearn.utils

In [ ]:
cc1=cc[cc['Class']==1]
cc0=cc[cc['Class']==0]

In [ ]:
cc1.shape

(213, 31)

In [ ]:
cc0.shape

(92967, 31)

In [ ]:
cc0_=resample(cc0,replace=False,n_samples=2000,random_state=42)
cc0_.shape

(2000, 31)

In [ ]:
data=pd.concat([cc0_,cc1])
data

Time        V1        V2        V3  ...       V27       V28   Amount  Class
8151   11037  1.226232 -0.109795  0.739753  ... -0.098373 -0.009103    15.95    0.0
26268  33976  0.990182 -0.053769  0.263690  ...  0.012580  0.035939   108.00    0.0
17312  28624  1.194535  0.211683  0.151196  ... -0.014067  0.012653     0.99    0.0
9710   14469  1.101430 -0.162202  0.927627  ... -0.080475 -0.001598    39.00    0.0
85949  61010 -0.151171  0.376210  1.865200  ...  0.150310  0.040380    34.87    0.0
...      ...       ...       ...       ...  ...       ...       ...      ...    ...
88876  62330  1.140865  1.221317 -1.452955  ...  0.099942  0.122988     1.00    1.0
88897  62341 -5.267760  2.506719 -5.290925  ... -1.486766  0.677664     1.10    1.0
89190  62467 -5.344665 -0.285760 -3.835616  ...  1.520613 -1.115937  1402.16    1.0
91671  63578 -0.639191 -0.085595  1.265452  ... -0.061962 -0.062687   290.18    1.0
92777  64093 -6.133987  2.941499 -5.593986  ... -1.684186  0.310195   294.90    1.0

[2213 rows x 31 columns]

In [ ]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
x=data.drop('Class',axis=1)
y=data['Class']


In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=42)

In [ ]:
from sklearn.preprocessing import StandardScaler
s=StandardScaler()
x_train=s.fit_transform(x_train)
x_test=s.fit_transform(x_test)

In [ ]:
model=GaussianNB()
model.fit(x_train,y_train)

GaussianNB()

In [ ]:
ypred=model.predict(x_test)

In [ ]:
ypred

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 1., 0., 0., 1., 1., 0., 1., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
       0., 1., 1., 0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

In [ ]:
accuracy_score(y_test,ypred)

0.9675090252707581